globus_sdk/config/environments.py

```diff
@@ -0,0 +1,85 @@
+import logging
+import typing
+
+log = logging.getLogger(__name__)
+
+
+class EnvConfig:
+    envname: str
+    domain: str
+    no_dotapi: typing.List[str] = ["auth"]
+
+    # this same dict is inherited (and therefore shared!) by all subclasses
+    _registry: typing.Dict[str, typing.Type["EnvConfig"]] = {}
+
+    # this is an easier hook to use than metaclass definition -- register every subclass
+    # in this dict automatically
+    #
+    # as a result, anyone can define
+    #
+    #       class BetaEnv(EnvConfig):
+    #           domain = "beta.foo.bar.example.com"
+    #           envname = "beta"
+    #
+    # and retrieve it with get_config_by_name("beta")
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._registry[cls.envname] = cls
+
+    @classmethod
+    def get_service_url(cls, service):
+        # you can override any name with a config attribute
+        service_url_attr = f"{service}_url"
+        if hasattr(cls, service_url_attr):
+            return getattr(cls, service_url_attr)
+
+        # the typical pattern for a service hostname is X.api.Y
+        # X=transfer, Y=preview.globus.org => transfer.api.preview.globus.org
+        # check `no_dotapi` for services which don't have `.api` in their names
+        if service in cls.no_dotapi:
+            return f"https://{service}.{cls.domain}/"
+        return f"https://{service}.api.{cls.domain}/"
+
+
+# the main interfaces to this module
+#   get_config_by_name("test") -> TestEnvConfig
+def get_config_by_name(envname: str) -> typing.Optional[typing.Type[EnvConfig]]:
+    return EnvConfig._registry.get(envname)
+
+
+def get_service_url(environment, service):
+    log.debug(f'Service URL Lookup for "{service}" under env "{environment}"')
+    conf = get_config_by_name(environment)
+    if not conf:
+        raise ValueError(f'Unrecognized environment "{environment}"')
+    url = conf.get_service_url(service)
+    log.debug(f'Service URL Lookup Result: "{service}" is at "{url}"')
+    return url
+
+
+#
+# public environments
```

Comment:
Probably a matter of opinion, but maybe it would be worth separating the code & data here: if the following are meant to be static configuration, then I would probably put them in a separate file from the `EnvConfig` class itself.

Reply:
Yeah, I don't generally disagree. The only issue is that with the registry pattern, you need to be careful and ensure that when `EnvConfig` is imported, the data module is imported as well.

It's a matter of taste, but I think I prefer putting them together in one module (because the amount of data is quite small) to having to do something like a special import in `__init__.py`. If you have a particular rewrite in mind, I'm open to looking at the options though.